### PR TITLE
[futures] Add collectAny() to ivalue::Future for completeness

### PR DESCRIPTION
**Pull request by jjlilley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39597 [futures] Add collectAny() to ivalue::Future for completeness**

To complement collectAll(), this change adds collectAny(), and writes
up relevant unittest coverage.

We also remove the vector-based helper version of collectAll(), which
was debatable usefulness in a previous change.

Differential Revision: [D21910311](https://our.internmc.facebook.com/intern/diff/D21910311/)

### CODE REVIEW

File: aten/src/ATen/core/ivalue.cpp [hunk 1]
```diff
@@ -634,21 +634,47 @@ CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
   return ctx->dstFuture;
 }
 
-CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
-    std::vector<intrusive_ptr<ivalue::Future>> srcs) {
-  auto typePtr = srcs.empty() ? AnyType::get() : srcs[0]->elementType();
-  // Lists are generally expected to have a consistent type, check
-  // for this unless we have a compelling target use case.
-  for (size_t i = 1, len = srcs.size(); i < len; ++i) {
```
Comment (wanchaol) [inline aten/src/ATen/core/ivalue.cpp:642]: I think we should still those torch check logics in the `collectAll` c10::List overload, to ensure the single type list constraint.

File: aten/src/ATen/core/ivalue.cpp [hunk 2]
```diff
@@ -634,21 +634,47 @@ CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
...
-  List<intrusive_ptr<ivalue::Future>> asList(FutureType::create(typePtr));
-  asList.reserve(srcs.size());
-  for (auto&& s : srcs) {
-    asList.push_back(std::move(s));
+  TypePtr typePtr = srcs.get(0)->elementType();
+  for (size_t i = 0, tot = srcs.size(); i < tot; ++i) {
+    if (srcs.get(i)->completed()) {
+      return srcs.get(i);
+    }
+    TORCH_CHECK(i == 0 || (*typePtr == *srcs.get(i)->elementType()));
```
Comment (wanchaol) [inline aten/src/ATen/core/ivalue.cpp:649]: nit: we don't need i==0 here?

**Review by wanchaol — APPROVED** [review]
looks good except some minor comments.

File: aten/src/ATen/core/ivalue.cpp (hunk 2 above)
Comment (jjlilley) [reply aten/src/ATen/core/ivalue.cpp:649]: The `i==0` is to skip the (more expensive) type check on the right for the i ==0 case

File: aten/src/ATen/core/ivalue.cpp (hunk 1 above)
Comment (jjlilley) [reply aten/src/ATen/core/ivalue.cpp:642]: Sounds good,  adding a `TORCH_CHECK` to the `addCallback` loop.

**Pull request closed without merging**